### PR TITLE
[WIP] Introduce dev tool: services graph visualizer

### PR DESCRIPTION
**Pull request by matks**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Introduce a dev tool able to extract and display the Symfony services graph being used by a controller
| Type?         | new feature
| Category?     | CO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | 
| How to test?  | New feature only available for devs. But you can check it at `/admin-dev/index.php/dev/services-graph`

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/11612)

## Screenshot
capture d ecran 2018-12-04 a 18 42 44

### CODE REVIEW

**Comment by matks** [comment]
## Known issues

- https://github.com/PrestaShop/PrestaShop/pull/11612/files#diff-be8eb857f87cfff8499ff11beeaa2f35R35 hard dependency to http://www.webgraphviz.com/viz.js JS lib
- https://github.com/PrestaShop/PrestaShop/pull/11612/files#diff-bcd56a42c9b4c0f2ad64305ba5941271R40 CustomGraphvizDumper is a copy-paste of https://github.com/symfony/dependency-injection/blob/master/Dumper/GraphvizDumper.php just for this: https://github.com/PrestaShop/PrestaShop/pull/11612/files#diff-bcd56a42c9b4c0f2ad64305ba5941271R168
- only works with `$this->get('...');` statements in the controller https://github.com/PrestaShop/PrestaShop/pull/11612/files#diff-4563d0abb690a677bbe22ed84956cfa2R72

## Can be improved

- only target a given Controller action instead of the whole Controller
- better display (colors, sizing)
- better handling of not-services dependencies https://github.com/PrestaShop/PrestaShop/pull/11612/files#diff-4563d0abb690a677bbe22ed84956cfa2R158
- files should be located in a better folder / with a better namespace
- URL should be chosen with other Core Team members (for all dev tools)
- add documentation links to computed HTMl graph

File: src/PrestaShopBundle/Resources/views/Dev/graph.html.twig [hunk 1]
```diff
@@ -0,0 +1,55 @@
+<html xmlns="http://www.w3.org/1999/html">
+<head>
+  <meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
+  <meta charset="utf-8">
+  <title>PrestaShop service visualisation using graphviz</title>
+</head>
+<body>
+
+<div>
```
Comment (mickaelandrieu) [inline src/PrestaShopBundle/Resources/views/Dev/graph.html.twig:9]: can we use a Symfony form for that?

File: src/PrestaShopBundle/Controller/Dev/ServicesGraphDumper.php [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+
+        if ((is_string($argument)) || (is_array($argument))) {
+            return;
+        }
+        if (($argument instanceof Parameter) || ($argument instanceof \Symfony\Component\ExpressionLanguage\Expression)) {
+            return;
+        }
+
+        if ($argument instanceof Reference) {
+            $serviceId = (string)$argument;
```
Comment (mickaelandrieu) [inline src/PrestaShopBundle/Controller/Dev/ServicesGraphDumper.php]: you don't need to cast, the id is already casted in Reference constructor

File: src/PrestaShopBundle/Controller/Dev/ServicesGraphVisualizerController.php [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to http://www.prestashop.com for more information.
+ *
+ * @author    PrestaShop SA <contact@prestashop.com>
+ * @copyright 2007-2018 PrestaShop SA
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
+namespace PrestaShopBundle\Controller\Dev;
```
Comment (mickaelandrieu) [inline src/PrestaShopBundle/Controller/Dev/ServicesGraphVisualizerController.php:27]: thinking about it, if it's only available for people connected into the back office: it needs to be put into `Admin` namespace. How about `PrestaShopBundle\Controller\Admin\Dev` ?

File: src/PrestaShopBundle/Controller/Dev/ServicesGraphVisualizerController.php [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+use Symfony\Component\DependencyInjection\Definition;
+use Symfony\Component\DependencyInjection\Exception\ServiceNotFoundException;
+use Symfony\Component\DependencyInjection\Loader\YamlFileLoader;
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+use Symfony\Component\DependencyInjection\Parameter;
+use Symfony\Component\DependencyInjection\Reference;
+use Symfony\Component\HttpFoundation\Request;
+
+
+class ServicesGraphVisualizerController extends Controller
```
Comment (mickaelandrieu) [inline src/PrestaShopBundle/Controller/Dev/ServicesGraphVisualizerController.php:33]: needs some love regarding CS and PHPDocs

**Review by mickaelandrieu — COMMENTED** [review]
Great feature! I have a lot of ideas to enrich and use it.

Good job @matks !

File: src/Core/Dev/ServicesGraph/ServicesGraphDumper.php [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+     * (recursive search)
+     *
+     * @param string[] $rootNodes service IDS
+     *
+     * @return ContainerBuilder
+     */
+    protected function buildServiceContainerFromRootNodes(array $rootNodes)
+    {
+        $containerBuilder = $this->getPrestaShopContainerBuilder();
+        $this->containerBuilder = $containerBuilder;
```
Comment (jolelievre) [inline src/Core/Dev/ServicesGraph/ServicesGraphDumper.php:125]: why not directly:

```
$this->containerBuilder = this->getPrestaShopContainerBuilder();
```

File: src/Core/Dev/ServicesGraph/ServicesGraphDumper.php [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+
+        if ($argument instanceof Reference) {
+            $serviceId = (string)$argument;
+            try {
+                $definition = $this->containerBuilder->getDefinition($serviceId);
+            } catch (ServiceNotFoundException $e) {
+                // @todo: to be improved: this could be part of the graph
+                return;
+            }
+            $this->tagDefinitionRecursivelyForGraphExploration($definition, $i);
```
Comment (jolelievre) [inline src/Core/Dev/ServicesGraph/ServicesGraphDumper.php:217]: the tag is added twice: in `tagDefinitionRecursivelyForGraphExploration` and at the end of `tagDefinitionArgumentRecursively` (should be handled inside the `else`)
also is it logic to tag a service if it hasn't been found?

**Review by jolelievre — CHANGES_REQUESTED** [review]
this is awesome! this can really help for the migration if you look at a page as an example

extra idea: we could add a link between this feature and the symfony dev toolbar
https://symfony.com/doc/3.4/profiler/data_collector.html

**Comment by matks** [comment]
Attempted to use main PS layout for the page, but this POC uses a higher jquery version 😭. Workaround that might actually turn even better: put this graph into the SF debug bar.

In SF debug bar, I can map what controller & action has been used. It should make the parsing even easier.

**Comment by matks** [comment]
I cant find time to continue this project. I close this PR and might make it a side-project instead.

**Pull request closed without merging**